httpbp/handler.go

```diff
@@ -0,0 +1,74 @@
+package httpbp
+
+import (
+	"context"
+	"html/template"
+	"net/http"
+)
+
+type HandlerFunc func(context.Context, *http.Request, Response) (interface{}, error)
```

Comment:
A few design discussions regarding this type:

1. Currently it could return an error, is that supposed to be `HTTPError`, or is that not supposed to be `HTTPError` and the handler will wrap whatever error it returns into an `HTTPError`?
2. If it's `HTTPError`, since `Response` is a subset of `HTTPError`, should the implementation still try to write to Response? Or is it that in case of error, don't touch Response, return an `HTTPError` instead?

Reply:
1. It can be either, if you return an HTTPError, it uses that to formulate the error response, if it's another error, it just returns a generic `500` error (it doesn't actually use HTTPError for this, although it could, it just uses `http.Error`).
2. As it currently stands, HTTPError is separate from Response, so if you want to return an HTTP error you would return `nil, NewJSONError(...)` and wouldn't mess around with `Response`.  You _can_ do that but it won't do anything, the error will determine what is returned.